File: codes/Garnetwzy/377.cpp
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        vector<int> dp(target+1, -1);
        dp[0] = 1;
        return help(nums, dp, target);
    }
    
    int help(vector<int>& nums, vector<int>& dp, int target) {
        if(dp[target] != -1)
            return dp[target];
        int ret = 0;
        for(int num: nums) {
            if(num <= target) {
                ret += help(nums, dp, target-num);
            }
        }
        dp[target] = ret;
        return ret;
    }
    
};

File: codes/HellDream/377.java
class Solution {
    public int combinationSum4(int[] nums, int target) {
        if(nums.length==0||target<0) return 0;
        int[] targets = new int[target+1];
        targets[0] = 1;
        for(int i=1;i<=target;i++){
            for(int j=0;j<nums.length;j++){
                if(i>=nums[j]){
                    targets[i] += targets[i-nums[j]];
                }
            }
        }
        return targets[target];
    }
}